### ISSUE TITLE
[Bug] DeepSeek-V4 checkpoint with fp8 routed experts scores 10pp lower on AIME25 than vLLM (same weights, same node, same harness)

### ISSUE THREAD

**Issue by danielliu99** [sgl-project/sglang#39311]
### Checklist

- [x] I searched related issues but found no solution.
- [ ] **The bug persists in the latest version.** — See "Version note" below. This was measured on **v0.5.17**, not the latest v0.5.19. v0.5.19 requires `torch==2.13.0` while this deployment is pinned to `torch 2.11.0+cu130` to stay aligned with the training stack; I did not want to report numbers from an unsupported torch/sglang combination. Happy to re-run on v0.5.19 if you can confirm the expected torch/flashinfer pairing.
- [x] Environment info and reproduction details are included below.
- [x] This is a bug report, not a general question.
- [x] Written in English.

### Describe the bug

On the **same machine, same weights, same harness, same sampling parameters, same day**, SGLang scores **10.2 points lower than vLLM on AIME25** for a DeepSeek-V4-Flash checkpoint whose routed experts are **fp8**.

The official base model (fp4 routed experts) does **not** show this gap — it scores 95.83% under SGLang, in line with the ~96 reported in the SGLang cookbook. The gap appears only with the fp8-expert variant.

Runs were strictly sequential: SGLang was stopped and GPU memory fully released before vLLM was started, so the two never shared the GPUs.

#### Results — `sgl-eval run aime25`, 30 problems × 16 repeats = 480 samples

| | **SGLang 0.5.17** | **vLLM 0.28.0** |
|---|---|---|
| **pass@1[avg-of-16]** | **84.58%** ± 6.87 (SEM 1.72) | **94.79%** ± 4.21 (SEM 1.05) |
| majority@16 | 96.67% | 100.00% |
| pass@16 | 100.00% | 100.00% |
| **no_answer** | **4.17%** | **0.83%** |
| truncated_rate | 0.00% | 0.00% |
| error_rate | 0.00% | 0.00% |

The 10.2pp gap is far outside the SEM of either run.

#### The gap is mostly wrong answers, not lost answers

A natural first guess is that the parser is eating answers. Recomputing from the dumped predictions after excluding every `no_answer` case:

| | correct / answered | accuracy |
|---|---|---|
| SGLang | 406 / 459 | **88.45%** |
| vLLM | 455 / 476 | **95.59%** |

Still a **7.1pp** gap. So `no_answer` accounts for only about a third of the difference — the majority is the model producing *wrong* answers under SGLang. That points at the generation path rather than the parsing layer.

`truncated_rate` is 0.00% on both sides, so this is not a `max_tokens` artifact.

### The distinguishing factor: fp8 routed experts

This is the same model family and the same SGLang version that works fine on the official base weights. The only difference is the expert quantization:

| | This checkpoint (gap) | Official base (no gap) |
|---|---|---|
| `expert_dtype` | **`"fp8"`** | `"fp4"` |
| `quantization_config.scale_fmt` | **`"float"`** (F32 block scale) | `"ue8m0"` |
| `quant_method` | `fp8` | `fp8` |
| `weight_block_size` | `[128, 128]` | `[128, 128]` |

Tensor names and shapes are identical between the two; they differ only in expert dtype and scale dtype (verified by a header-by-header comparison).

SGLang detects this correctly at startup:

```
Auto-detected DSV4 routed-expert layout: is_fp4_experts=False
```

and routes to `Fp8MoEMethod` + DeepGEMM, which is the expected path. Loading succeeds, there are no warnings, and throughput is normal — the output is simply less accurate.

**The weights are from an internal fine-tune and cannot be shared**, but they were produced by `ms-swift`'s `megatron export` with `fp8_recipe=blockwise` + `fp8_param_gather`, which yields exactly the `expert_dtype=fp8` / `scale_fmt=float` combination above. Any DeepSeek-V4 checkpoint exported that way should reproduce this. If it would help, I can run additional diagnostics on this checkpoint on request (e.g. greedy decoding traces, per-layer activation dumps).

### Reproduction

Both servers were started on the same 8×H200 node, one at a time.

**SGLang** (no `--moe-runner-backend marlin` — marlin MoE only engages when `is_fp4_experts` is true; passing it with fp8 experts fails with `no attribute 'runner'`):

```bash
python -m sglang.launch_server \
  --model-path /path/to/DeepSeek-V4-Flash-SFT \
  --served-model-name DeepSeek-V4-Flash-SFT \
  --host 0.0.0.0 --port 8001 \
  --tp-size 8 --trust-remote-code \
  --reasoning-parser deepseek-v4 --tool-call-parser deepseekv4 \
  --mem-fraction-static 0.8 --watchdog-timeout 1800 --enable-metrics
```

**vLLM** (official recipe, no extra env vars):

```bash
vllm serve /path/to/DeepSeek-V4-Flash-SFT \
  --served-model-name DeepSeek-V4-Flash-SFT \
  --trust-remote-code --kv-cache-dtype fp8 --block-size 256 \
  --tensor-parallel-size 8 --no-enable-flashinfer-autotune \
  --tokenizer-mode deepseek_v4 --tool-call-parser deepseek_v4 \
  --enable-auto-tool-choice --reasoning-parser deepseek_v4 \
  --reasoning-config '{"reasoning_parser":"deepseek_v4","reasoning_start_str":"","reasoning_end_str":""}' \
  --host 0.0.0.0 --port 8001
```

**Evaluation** — identical against both endpoints:

```bash
sgl-eval run aime25 \
  --base-url http://127.0.0.1:8001/v1 \
  --model DeepSeek-V4-Flash-SFT --api-key EMPTY \
  --n-repeats 16 --max-tokens 200000 \
  --temperature 1.0 --top-p 1.0 --thinking \
  --out-dir ./aime_out
```

### What I ruled out

Listing these so nobody re-treads the same ground:

1. **Chat template injecting tool scaffolding.** The model directory ships no `chat_template`, so each engine uses its own built-in. Measured `usage.prompt_tokens` on SGLang for a single-message request: **5 tokens** without `tools` vs **276** with one tool defined (the message itself is 1 token). So the no-tools template overhead is ~4 tokens — there is no hidden tool scaffolding in the prompt.
2. **`--moe-runner-backend marlin`.** Not applicable to fp8 experts (see above); auto-detection is doing the right thing.
3. **Truncation.** `truncated_rate` is 0.00% on both engines at `--max-tokens 200000`.
4. **`num_reasoning_tokens` reporting.** In the dumped predictions this is 0 for 100% of SGLang samples vs 0.21% for vLLM, which looks alarming but is unrelated: `sgl-eval` reads `usage.completion_tokens_details.reasoning_tokens` (`sampler.py:172`), which SGLang does not populate. The reasoning content itself is still returned in `message.reasoning_content`.

### Environment

`python3 -m sglang.check_env` (version section; full NVIDIA topology omitted for length — 8×H200, NV18 all-to-all, 2 NUMA nodes):

```
Python: 3.12.13 | packaged by Anaconda, Inc. | (main, Jul  9 2026, 14:38:16) [GCC 14.3.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H200
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.9, V12.9.86
CUDA Driver Version: 590.48.01
PyTorch: 2.11.0+cu130
sglang: 0.5.17
sglang-kernel: 0.4.5
flashinfer_python: 0.6.15.post1
flashinfer_cubin: Module Not Found
flashinfer_jit_cache: Module Not Found
triton: 3.6.0
transformers: 5.12.1
torchao: 0.17.0
numpy: 2.3.5
huggingface_hub: 1.27.0
outlines: 0.1.11
xgrammar: 0.2.1
openai: 2.6.1
torchcodec: 0.11.1
ulimit soft: 1024
```

Comparison engine: **vLLM 0.28.0**, same node, same weights.

### Version note

Measured on v0.5.17. v0.5.19 declares `torch==2.13.0` and `flashinfer_python[cu13]==0.6.18`; this deployment is pinned to `torch 2.11.0+cu130` to match the training stack, and I did not want to report numbers from a hand-forced, unsupported combination. If you can confirm the intended torch/flashinfer pairing for v0.5.19 on H200, I'm glad to re-run the same comparison and post the numbers here.
